plugins/plugins.go

```diff
@@ -72,17 +89,93 @@ func (m *Manager) Register(plugin Plugin) {
 	m.plugins = append(m.plugins, plugin)
 }
 
+// GetCompiler returns the compiler instance
+func (m *Manager) GetCompiler() *ast.Compiler {
+	m.mtx.RLock()
+	defer m.mtx.RUnlock()
+	return m.Compiler
+}
+
+func (m *Manager) setCompiler(compiler *ast.Compiler) {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	m.Compiler = compiler
+}
+
+// RegisterCompilerTrigger registers for change notifications
+// when the compiler is changed.
+func RegisterCompilerTrigger(config CompilerTriggerConfig) {
+	registeredTriggersMux.Lock()
+	defer registeredTriggersMux.Unlock()
+	registeredTriggers = append(registeredTriggers, config)
+}
+
 // Start starts the manager.
 func (m *Manager) Start(ctx context.Context) error {
 	if m == nil {
 		return nil
 	}
+
+	err := storage.Txn(ctx, m.Store, storage.TransactionParams{}, func(txn storage.Transaction) error {
+		// load policies from storage
+		compiler, err := loadCompilerFromStore(ctx, m.Store, txn)
+		if err != nil {
+			return err
+		}
+		// set compiler on manager
+		m.setCompiler(compiler)
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	// start plugins
 	for _, p := range m.plugins {
 		if err := p.Start(ctx); err != nil {
 			return err
 		}
 	}
-	return nil
+
+	return storage.Txn(ctx, m.Store, storage.WriteParams, func(txn storage.Transaction) error {
+		_, err := m.Store.Register(ctx, txn, storage.TriggerConfig{OnCommit: func(ctx context.Context, txn storage.Transaction, event storage.TriggerEvent) {
+			if event.PolicyChanged() {
```

Comment:
For clarity, I would factor the inner function into a private function on the manager.

Reply:
Moved inner into a private manager func